### ISSUE TITLE
Provide filtered export of envoy metrics to replace statsd-to-prometheus function

### ISSUE THREAD

**Issue by douglas-reid** [istio/istio#8415]
**Describe the feature request**
Many of the metrics about envoy's operation (distinct from request/response metrics) are held exclusively by envoy. To date, we've been using a statsd-to-prometheus process to expose those metrics for consumption by prometheus, etc. Unfortunately, this does not scale well (as there is no filtering) and does not properly handle aggregation across envoy instances.

In this FR, I propose augmenting the envoy metrics consumption to provide a limited set of metrics for collection and to collect them in a more scalable, robust fashion.

**Describe alternatives you've considered**
There are numerous ways to improve the statsd-to-prometheus experience (providing regex-based filters and using per-pod deployment), as well as ways in which overall telemetry can be improved. Another option would be to drop envoy metrics collection altogether. But, I think ultimately, we need *some* way to provide filtered, high-value metrics about the proxies to operators.

**Additional context**
I will be working on a doc on this FR for presentation and discussion in the policies-and-telemetry WG. I welcome all input and contributions -- especially in identifying an initial set of high value metrics.

cc: @costinm @Stono @kyessenov

**Comment by kyessenov** [comment]
Summary of preliminary discussion.

Mechanics:
- statsd bridge proves to be a performance bottleneck
- we can use gRPC metrics service to pipe out envoy metrics

Envoy metrics:
- version summary per envoy instance x envoy resource
- envoy memory use

Application metrics:
- annotate request metrics with envoy/istio resource metadata

**Comment by costinm** [comment]
That's pretty much not what we discussed...

Statsd bridge - with the fix to label IP - is not a performance bottleneck, and with a config change to filter the metrics will be even less so.  Using gRPC metrics instead of UDP would only make things far worse in terms of scalability - I would like to have a load tested prototype compared before we waste time debating this. 

The envoy metrics prove to be _very_ useful, and we'll try to identify a list that makes sense for 
pilot and config distribution - but regardless of what design will be done for future, in 1.0 we need to 
fix the configs and make it usable. I suppose this can remain a baseline to compare any new designs against.

**Comment by costinm** [comment]
To be clear: statsd bridge is stable and while it uses lots of memory - it is still less than telemetry or prometheus itself, and appears more stable than both. The original problem was that it didn't label the data ( so metrics were meaningless ), and after it was fixed the alloc had to be adjusted. I suspect after filtering it'll work even better. 

Even when it was miss-configured and crashing - it resulted in transient metric loss but no cascading error ( UDP just got dropped ).

**Comment by costinm** [comment]
@nmittler  is merging the readiness probe for pilot-agent, which is scrapping metrics from envoy. Once it is done, we'll also have the option to expose envoy directly for scrapping ( this will be in 1.0 - so can be made available soon to prod users ). It's another short term option until a new design gets implemented, and would reduce the CPU cost added by the UDP bridge.

**Comment by kyessenov** [comment]
I just read the code, and it's only reading the number of inbound clusters and listeners. Is that all we need? What do you mean by exposing envoy directly to scraping? Do you mean to aggregate prom metrics in the agent? Why not just scrape directly from envoy (it _has_ prom backend)? 

I find your comment mysterious. Do you mind starting a doc with more explanation?

**Comment by douglas-reid** [comment]
Proposal doc: https://tinyurl.com/y7tc5ohd.

**Comment by douglas-reid** [comment]
Closing this issue, as we have a working approach in 1.0.3. We can open a separate issue for tracking modifications / improvements.

**Comment by mhamrah** [comment]
Can you summarize the working approach? The proposal doc seems to be protected. Thank you.

**Comment by douglas-reid** [comment]
@mhamrah the doc should be available via the Istio community folder on drive. If you are still having issues accessing it, please reach out again to confirm.

The approach currently used is as follows:
- custom tag configuration in envoy bootstrap for prometheus metrics (to adjust for some errors in the default + customization for Istio cluster names)
- a special port exposed for sidecars that allows access to the prometheus endpoint on each envoy
- prometheus config to look for pods with the endpoint exposed and scrape
- prometheus config that filters out a large number of envoy metrics (restricted set for scalability, etc., concerns at the moment).

Hope that helps.
